src/util/schnorr.rs

```diff
@@ -75,9 +79,71 @@ impl TweakedPublicKey {
     pub fn new(key: PublicKey) -> TweakedPublicKey {
         TweakedPublicKey(key)
     }
-    
-    /// Returns the underlying public key 
+
+    /// Returns the underlying public key
     pub fn into_inner(self) -> PublicKey {
         self.0
     }
+}
+
+/// A BIP340-341 serialized schnorr signature with the corresponding hash type.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub struct SchnorrSig {
+    /// The underlying schnorr signature
+    pub sig: secp256k1::schnorrsig::Signature,
+    /// The corresponding hash type
+    pub hash_ty: SchnorrSigHashType,
+}
+
+impl SchnorrSig {
+
+    /// Deserialize from slice
+    pub fn from_slice(sl: &[u8]) -> Result<Self, SchnorrSigError> {
+        match sl.split_last() {
+            Some((hash_ty, sig)) => {
+                let hash_ty = SchnorrSigHashType::from_u8(*hash_ty)
+                    .map_err(|_| SchnorrSigError::InvalidSighashType(*hash_ty))?;
+                let sig = secp256k1::schnorrsig::Signature::from_slice(sig)?;
+                Ok(SchnorrSig { sig, hash_ty })
+            },
+            None => {
+                Err(SchnorrSigError::Secp256k1(secp256k1::Error::InvalidSignature))?
+            },
+        }
+    }
+
+    /// Serialize SchnorrSig
+    pub fn serialize(&self) -> Vec<u8> {
+        // TODO: add support to serialize to a writer to SerializedSig
+        let mut ser_sig = self.sig.as_ref().to_vec();
+        ser_sig.push(self.hash_ty as u8);
+        ser_sig
+    }
+}
+
+/// A key-related error.
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
+pub enum SchnorrSigError {
```

Comment:
Are we sure that we need a dedicated error type for each of sigs?

Reply:
Didn't you imply it above? :)

Reply:
Above I meant dedicated error variants, not types. Sorry for spelling thad badly

Reply:
I personally like keeping error types separate so that impossible error cases don't need to be handled by the callers. But I do understand most people probably find it annoying, especially if there are many error types and they want to just return a common type. One of my ten billion project ideas is having a (proc) macro generate "subset" error types (so you can have one common `Error` for library and then other errors with variants that are strict subsets and convertible to the common `Error`)...

Reply:
I try to separate error types for the exact same reason. It is really hard to write code downstream that does anything apart from propagating the error forward.

Reply:
Yeah, I always have dedicated error types in my projects, unless all of the options for two error types are the same (like in this case).

Reply:
Ah, true, they have same variants. I think it'd make sense to unify them unless there's expectation they will change in the future. Seems unlikely?

Reply:
They are slightly different types. Because while parsing `EcdsaSig` we have NonStandardSigHashType, whereas while parsing `SchnorrSig` we have `InvalidSigHashType`. 

Types apart from standard 6 types automatically are treated as SIGHASH_ALL or `ALL|ACP`. But the point is 0x04 is a valid consensus ECDSA sighash type, but 0x04 is consensus invalid `SchnorrSighashType`. I had specifically named them `NonStandard` and `Invalid` for this reason.

Reply:
Ah, I agree with having them separated then.